src/lambda/config/templates.ts

```diff
@@ -0,0 +1,139 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+import { parse, ParseError, ParseErrorCode } from 'jsonc-parser'
+import * as os from 'os'
+import * as path from 'path'
+import * as vscode from 'vscode'
+import * as fsUtils from '../../shared/filesystemUtilities'
+import { getLogger, Logger } from '../../shared/logger'
+
+export interface TemplatesConfig {
+    templates: {
+        [relativePath: string]: TemplateConfig
+    }
+}
+
+export interface TemplateConfig {
+    parameterOverrides?: {
+        [key: string]: string
+    }
+}
+
+export interface LoadTemplatesConfigContext {
+    logger: Pick<Logger, 'warn'>
+    readFile(path: string): Promise<string>
+}
+
+function getWorkspaceFolderPath(input: Pick<vscode.WorkspaceFolder, 'uri'> | vscode.Uri | string): string {
+    if (typeof input === 'string') {
+        return input
+    }
+
+    const workspaceFolder = input as Pick<vscode.WorkspaceFolder, 'uri'>
+    if (workspaceFolder.uri) {
+        return workspaceFolder.uri.fsPath
+    }
+
+    return (input as vscode.Uri).fsPath
+}
+
+export function load(
+    workspaceFolder: Pick<vscode.WorkspaceFolder, 'uri'>,
+    context?: LoadTemplatesConfigContext
+): Promise<TemplatesConfig>
+export function load(
+    workspaceUri: vscode.Uri,
+    context?: LoadTemplatesConfigContext
+): Promise<TemplatesConfig>
+export function load(
+    workspacePath: string,
+    context?: LoadTemplatesConfigContext
+): Promise<TemplatesConfig>
+export async function load(
+    workspaceFolderOrUriOrPath: Pick<vscode.WorkspaceFolder, 'uri'> | vscode.Uri | string,
+    context: LoadTemplatesConfigContext = {
+        logger: getLogger(),
+        readFile: async p => await fsUtils.readFileAsString(p)
+    }
+): Promise<TemplatesConfig> {
+    const templatesConfigPath = path.join(
+        getWorkspaceFolderPath(workspaceFolderOrUriOrPath),
+        '.aws',
+        'templates.json'
+    )
+
+    try {
+        const raw = await context.readFile(templatesConfigPath)
+        const errors: ParseError[] = []
+        const config = parse(raw, errors) as TemplatesConfig
+        if (errors.length > 0) {
+            const message = errors.length === 1 ?
+                ` ${formatParseError(errors[0])}`  :
+                `${os.EOL}${errors.map(formatParseError).join(os.EOL)}`
+
+            throw new Error(`Could not parse .aws/templates.json:${message}`)
+        }
+
+        return config
+    } catch (err) {
+        if (Array.isArray(err) && (err as any[]).length === 1) {
+            err = (err as any[])[0]
+        }
+
+        context.logger.warn(`Could not load .aws/templates.json: ${err}`)
+
+        return {
```

Comment:
I wonder if we should throw, so the calling code has a chance to decide what to do. We risk stomping the file _if_ we will have code that modifies and saves the file back out.

Reply:
Now that you mention it, there are several changes we should make in this area:

1. Check if the file is open in a `vscode.TextEditor`, and read contents from the editor if it is (in case the editor contains unsaved changes).
2. Distinguish between 'the file does not exist' and other access errors. In the former case, this is expected and we don't need to throw. In the latter, we can throw.
3. Add a corresponding `save` method.